### PR TITLE
Proxy deletion related fixes

### PR DESCRIPTION
**Pull request by aurimasc**
Proxy-related changes:
- drop automatic deletion of WeakProxy on finalize() (from now you have to manually manage it's lifetime); this is a very bad idea for few reasons:
  - it uses delete operator to destroy the proxy, so if you create C++ object in other way than operator new, you have to ensure C++ object is destroyed before Java object is finalized, otherwise crash
  - with C++ object having a weak reference to Java object, you have no way of controlling when the later one gets GCed
  - found 2 cases where we use WeakProxy in Unity code, one correctly uses new, other does not; the fact that you pretty much must use new is not evident, hence disaster waiting to happen
- ProxyTracker::DeleteAllProxies() used to delete each proxy while under mutex lock, where destructor of proxy would unregister itself leading to deadlock; proxies no longer deleted, only disabled, so problem solved
- Proxies are no longer tracket genericly; after struggling with making unity tests pass, I conclude that this is a bad thing in general. All proxy object should be properly destroyed instead of having such generic destruction that it is not clear when it's safe to call (should be called after proxies are guaranteed to no longer serve a useful purpose, yet the object still alive in valid memory)

Not related changes:
- Fixed a couple of warnings in test project
- Switching codegen to debug will build library with debug symbols (useful for local)

Katana green (couple proxies temporarily commented out in unity, cause they are known leaks to fix).

### CODE REVIEW

File: APIHelper.h [hunk 1]
```diff
@@ -82,6 +82,13 @@ class Ref
 		return *this;
 	}
 
+	void Release()
+	{
+		if (m_Ref && !m_Ref->Release())
```
Comment (todi1856) [inline APIHelper.h:88]: I know this an old code, but it seems we don't validate if Release is called too much. Meaning the counter might become negative, making this m_Ref->Release() return true. Should we care, put an assert or something?

**Review by todi1856 — APPROVED** [review]
minor comment

File: APIHelper.h (hunk 1 above)
Comment (aurimasc) [reply APIHelper.h:88]: We null the m_Ref after each release (now, after changes), so it's self protected from excessive releases. Previously it was protected from that being private.

File: test/Test.cpp [hunk 2]
```diff
@@ -231,7 +231,7 @@ int main(int argc, char** argv)
 		java::lang::Integer integers4[] = { 1, 2, 3, 4 };
 		jni::Array<jint> test05(4, integers4);
 		for (int i = 0; i < test05.Length(); ++i)
-			printf("ArrayTest05[%ld],", (jint)test05[i]);
+			printf("ArrayTest05[%d],", (int)test05[i]);
```
Comment (juliusmik) [inline test/Test.cpp:244]: why is this needed?

**Review by juliusmik — APPROVED** [review]

File: test/Test.cpp (hunk 2 above)
Comment (aurimasc) [reply test/Test.cpp:244]: On Windows jint is typedefed as long, while on Mac as int. So the original code compiles fine on Windows, but gives warnings on Mac, since %ld means long, but is given int.

File: APIHelper.h (hunk 1 above)
Comment (todi1856) [reply APIHelper.h:88]: hmph, okay

**Comment by todi1856** [comment]
I get the idea about the whole - let's manually delete proxies, but I think now we don't have any safeguards against leaks. Maybe refactor ProxyTracker to at least report leaks on quit, similarly what we have with Unity's memory manager, where during runtime shutdown, it prints if we're leaking.

File: Proxy.h [hunk 3]
```diff
@@ -1,20 +1,15 @@
...
-class ProxyTracker;
-
 class ProxyObject : public virtual ProxyInvoker
 {
 // Dispatch invoke calls
 public:
 	virtual jobject __Invoke(jclass clazz, jmethodID mid, jobjectArray args);
-// Cleanup all proxy objects
-	static void DeleteAllProxies();
+	virtual void DisableProxy() = 0;
```
Comment (todi1856) [inline Proxy.h:22]: can you add coverge for this new function ? can i call DisableProxy twice?

**Review by todi1856 — COMMENTED** [review]
new changes reviewed. questions

**Comment by aurimasc** [comment]
> I get the idea about the whole - let's manually delete proxies, but I think now we don't have any safeguards against leaks. Maybe refactor ProxyTracker to at least report leaks on quit, similarly what we have with Unity's memory manager, where during runtime shutdown, it prints if we're leaking.

If you create proxies dynamically, they are tracked by memory manager (in unity). Those that are not created dynamically are automatically destroyed, so no leak either.
So the tracking of leaks seems to be only relevant if this lib is used outside of Unity, which for one doesn't exist (to my knowledge), for other I don't think we should worry about it, it's users responsibility.
Unless I miss some use case.

File: Proxy.h (hunk 3 above)
Comment (aurimasc) [reply Proxy.h:22]: Good one, it revealed that I misplaced it and it was protected :)

**Comment by todi1856** [comment]
> > I get the idea about the whole - let's manually delete proxies, but I think now we don't have any safeguards against leaks. Maybe refactor ProxyTracker to at least report leaks on quit, similarly what we have with Unity's memory manager, where during runtime shutdown, it prints if we're leaking.
> 
> If you create proxies dynamically, they are tracked by memory manager (in unity). Those that are not created dynamically are automatically destroyed, so no leak either. So the tracking of leaks seems to be only relevant if this lib is used outside of Unity, which for one doesn't exist (to my knowledge), for other I don't think we should worry about it, it's users responsibility. Unless I miss some use case.

I am missing some info here, does Unity create proxies dynamically ? Do proxies get created when using UnityEngine.AndroidJNIModule API ?

I think leaked proxies are more important than regular leaks, because if I understand correctly, not deallocated proxy keeps a reference to java object ? 

Even today, if you'll enable memory manager to show leaks, it will print dozens of them, so it would be quite difficult to see proxy leaks there. 

Curious to here @juliusmik opinion on this

**Comment by aurimasc** [comment]
> I am missing some info here, does Unity create proxies dynamically ?

Proxy is created when you construct the object of type that is descendant of Proxy<Iface> (or WeakProxy, where Iface is the generated class representing Java interface). Unity code creates them dynamically via RuntimeStatic or just "new".

**Comment by todi1856** [comment]
> > I am missing some info here, does Unity create proxies dynamically ?
> 
> Proxy is created when you construct the object of type that is descendant of Proxy (or WeakProxy, where Iface is the generated class representing Java interface). Unity code creates them dynamically via RuntimeStatic or just "new".

Would call like AndroidJNIHelper.CreateJavaProxy create such proxy ? or is it a different kind of proxy?

**Review by todi1856 — APPROVED** [review]

**Comment by aurimasc** [comment]
Runtime and integration tests are green on katana using latest jnibridge and couple of places in unity commented out (to be fixed).

**Review by juliusmik — APPROVED** [review]
